pkg/internal/apis/certmanager/validation/plugins/approval.go

```diff
@@ -0,0 +1,297 @@
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package plugins
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	admissionv1 "k8s.io/api/admission/v1"
+	authzv1 "k8s.io/api/authorization/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	"k8s.io/client-go/discovery"
+	"k8s.io/client-go/kubernetes"
+	authzclient "k8s.io/client-go/kubernetes/typed/authorization/v1"
+
+	"github.com/jetstack/cert-manager/pkg/apis/certmanager"
+	cmapi "github.com/jetstack/cert-manager/pkg/apis/certmanager/v1"
+	internalcmapi "github.com/jetstack/cert-manager/pkg/internal/apis/certmanager"
+	"github.com/jetstack/cert-manager/pkg/internal/apis/certmanager/validation/util"
+)
+
+// approval is responsible for reviewing where users attempting to approve or
+// deny a CertificateRequest have sufficient permissions to do so.
+type approval struct {
+	scheme *runtime.Scheme
+
+	sarclient      authzclient.SubjectAccessReviewInterface
+	discoverclient discovery.DiscoveryInterface
+}
+
+type signerResource struct {
+	// signer resource name
+	name       string
+	group      string
+	namespaced bool
+
+	// name of the object for this signer
+	signerName       string
+	requestNamespace string
+}
+
+func newApproval(scheme *runtime.Scheme) *approval {
+	return &approval{
+		scheme: scheme,
+	}
+}
+
+func (a *approval) Init(client kubernetes.Interface) {
+	a.sarclient = client.AuthorizationV1().SubjectAccessReviews()
+	a.discoverclient = client.Discovery()
+}
+
+// Validate will review whether the client is able to approve or deny the given
+// request, if indeed they are attempting to. A SubjectAccessReview will be
+// performed if the client is attempting to approve/deny the request. An error
+// will be returned if the SubjectAccessReview fails, or if they do not have
+// permissions to perform the approval/denial. The request will also fail if
+// the referenced signer doesn't exist in this cluster.
+func (a *approval) Validate(req *admissionv1.AdmissionRequest, oldObj, obj runtime.Object) *field.Error {
+	// Only perform validation on UPDATE operations
+	if req.Operation != admissionv1.Update {
+		return nil
+	}
+
+	// Only Validate over CertificateRequest resources
+	if req.RequestKind.Group != certmanager.GroupName || req.RequestKind.Kind != cmapi.CertificateRequestKind {
+		return nil
+	}
+
+	// Error if the clients are not initialised
+	if a.sarclient == nil || a.discoverclient == nil {
+		return internalError(errors.New("approval validation not initialised"))
+	}
+
+	gvk := schema.GroupVersionKind{
+		Group:   req.RequestKind.Group,
+		Version: runtime.APIVersionInternal,
+		Kind:    req.RequestKind.Kind,
+	}
```

Comment:
What's `runtime.APIVersionInternal` ? I expected this to be a cert-manager API version.

Reply:
Maybe add a few comments explaining what's going on here and in the following `scheme.Convert` step.

Reply:
`runtime.APIVersionInternal` is the internal API Version of a resource. I guess we could create a cert-manager one, the only thing that's important is that it is consistent, and could never conflict with a client facing version.